File: core/store.py
"""SQLite store: sessions, stints, recommendations, and learned per-car confidence.

Learning loop: when a new stint arrives for a (car, track) we already have a
stint for, and the setup changed in the direction a prior recommendation
proposed, the recommendation is graded on measured outcome (fuel-corrected
pace + symptom severity). Grades roll into per-(car, symptom, knob) confidence
that reorders future proposals — the app literally learns each car as people
drive it.
"""
import json
import os
import sqlite3
import time

from .session import numeric_value

SCHEMA = """
CREATE TABLE IF NOT EXISTS stints (
    id INTEGER PRIMARY KEY,
    created REAL,
    file_name TEXT,
    driver TEXT,
    car TEXT,
    track TEXT,
    stint_num INTEGER,
    setup_json TEXT,
    analysis_json TEXT,
    findings_json TEXT
);
CREATE TABLE IF NOT EXISTS grades (
    id INTEGER PRIMARY KEY,
    created REAL,
    car TEXT,
    track TEXT,
    symptom TEXT,
    knob TEXT,
    direction INTEGER,
    followed INTEGER,
    pace_delta REAL,
    severity_delta REAL,
    win INTEGER,
    before_stint INTEGER,
    after_stint INTEGER
);
CREATE INDEX IF NOT EXISTS idx_stints_car_track ON stints(car, track);
CREATE INDEX IF NOT EXISTS idx_grades_car ON grades(car, symptom, knob);
"""


def db_path():
    base = os.path.join(os.environ.get('LOCALAPPDATA', os.path.expanduser('~')), 'SetupLab')
    os.makedirs(base, exist_ok=True)
    return os.path.join(base, 'setuplab.db')


def connect():
    con = sqlite3.connect(db_path())
    con.row_factory = sqlite3.Row
    con.executescript(SCHEMA)
    return con


def save_stint(con, file_name, meta, stint_num, setup, analysis, findings):
    cur = con.execute(
        'INSERT INTO stints (created, file_name, driver, car, track, stint_num, setup_json, analysis_json, findings_json) '
        'VALUES (?,?,?,?,?,?,?,?,?)',
        (time.time(), file_name, meta['driver'], meta['car'], meta['track'], stint_num,
         json.dumps(setup), json.dumps(analysis), json.dumps(findings)))
    con.commit()
    return cur.lastrowid


def prior_stint(con, car, track, before_id):
    row = con.execute(
        'SELECT * FROM stints WHERE car=? AND track=? AND id<? ORDER BY id DESC LIMIT 1',
        (car, track, before_id)).fetchone()
    return row


def _severity_map(findings):
    return {f['symptom']: f['severity'] for f in findings}


def grade_against_prior(con, new_id):
    """Grade the prior stint's recommendations using this stint's measured outcome."""
    new = con.execute('SELECT * FROM stints WHERE id=?', (new_id,)).fetchone()
    if not new:
        return []
    prev = prior_stint(con, new['car'], new['track'], new_id)
    if not prev:
        return []

    old_setup = json.loads(prev['setup_json'])
    new_setup = json.loads(new['setup_json'])
    old_a, new_a = json.loads(prev['analysis_json']), json.loads(new['analysis_json'])
    old_f, new_f = json.loads(prev['findings_json']), json.loads(new['findings_json'])
    old_sev, new_sev = _severity_map(old_f), _severity_map(new_f)

    pace_delta = new_a['median_lap'] - old_a['median_lap']   # negative = faster
    graded = []
    for f in old_f:
        for p in f['proposals']:
            followed = 0
            for key in p['keys']:
                ov, nv = numeric_value(old_setup.get(key, '')), numeric_value(new_setup.get(key, ''))
                if ov is None or nv is None or nv == ov:
                    continue
                if (nv - ov) * p['direction'] > 0:
                    followed = 1
            if not followed:
                continue
            sev_delta = new_sev.get(f['symptom'], 0.0) - f['severity']
            win = 1 if (pace_delta < -0.02 or sev_delta < -0.3) else 0
            con.execute(
                'INSERT INTO grades (created, car, track, symptom, knob, direction, followed, '
                'pace_delta, severity_delta, win, before_stint, after_stint) VALUES (?,?,?,?,?,?,?,?,?,?,?,?)',
                (time.time(), new['car'], new['track'], f['symptom'], p['knob'], p['direction'],
                 1, pace_delta, sev_delta, win, prev['id'], new_id))
            graded.append({'symptom': f['symptom'], 'knob': p['knob'],
                           'pace_delta': pace_delta, 'severity_delta': sev_delta, 'win': bool(win)})
    con.commit()
    return graded


def learned_confidence(con, car):
    """{(symptom, knob): {'wins': n, 'tries': n, 'conf': laplace-smoothed}} for a car."""
    rows = con.execute(
        'SELECT symptom, knob, SUM(win) w, COUNT(*) n FROM grades WHERE car=? GROUP BY symptom, knob',
        (car,)).fetchall()
    return {(r['symptom'], r['knob']): {'wins': r['w'], 'tries': r['n'],
                                        'conf': (r['w'] + 1) / (r['n'] + 2)}
            for r in rows}


def apply_learning(findings, conf):
    """Annotate + reorder proposals by learned confidence (untested knobs keep 0.5 prior)."""
    for f in findings:
        for p in f['proposals']:
            c = conf.get((f['symptom'], p['knob']))
            p['learned'] = {'conf': c['conf'], 'tries': c['tries'], 'wins': c['wins']} if c \
                else {'conf': 0.5, 'tries': 0, 'wins': 0}
        f['proposals'].sort(key=lambda p: -p['learned']['conf'])
    return findings


def recent_stints(con, limit=50):
    return [dict(r) for r in con.execute(
        'SELECT id, created, file_name, driver, car, track, stint_num FROM stints '
        'ORDER BY id DESC LIMIT ?', (limit,)).fetchall()]


def get_stint(con, sid):
    r = con.execute('SELECT * FROM stints WHERE id=?', (sid,)).fetchone()
    return dict(r) if r else None

File: docs/js/store.js
// Learning store — IndexedDB port of core/store.py. Same grading logic:
// when a new stint's setup moved in a direction a prior recommendation
// proposed, grade it on measured outcome; grades roll into per-(car,
// symptom, knob) confidence that reorders future proposals.

import { numericValue } from './session.js';

const DB_NAME = 'setuplab';
const DB_VER = 1;

function openDb() {
  return new Promise((resolve, reject) => {
    const req = indexedDB.open(DB_NAME, DB_VER);
    req.onupgradeneeded = () => {
      const db = req.result;
      if (!db.objectStoreNames.contains('stints')) {
        const st = db.createObjectStore('stints', { keyPath: 'id', autoIncrement: true });
        st.createIndex('car', 'car');
        st.createIndex('car_track', ['car', 'track']);
      }
      if (!db.objectStoreNames.contains('grades')) {
        const gr = db.createObjectStore('grades', { keyPath: 'id', autoIncrement: true });
        gr.createIndex('car', 'car');
      }
    };
    req.onsuccess = () => resolve(req.result);
    req.onerror = () => reject(req.error);
  });
}

function tx(db, store, mode, fn) {
  return new Promise((resolve, reject) => {
    const t = db.transaction(store, mode);
    const result = fn(t.objectStore(store));
    t.oncomplete = () => resolve(result.__value !== undefined ? result.__value : result);
    t.onerror = () => reject(t.error);
  });
}

function reqAsync(req) {
  return new Promise((resolve, reject) => {
    req.onsuccess = () => resolve(req.result);
    req.onerror = () => reject(req.error);
  });
}

export async function saveStint(fileName, meta, stintNum, setup, analysis, findings) {
  const db = await openDb();
  const rec = {
    created: Date.now() / 1000, file_name: fileName,
    driver: meta.driver, car: meta.car, track: meta.track,
    stint_num: stintNum, setup, analysis, findings,
  };
  const t = db.transaction('stints', 'readwrite');
  const id = await reqAsync(t.objectStore('stints').add(rec));
  db.close();
  return id;
}

async function allStints() {
  const db = await openDb();
  const rows = await reqAsync(db.transaction('stints').objectStore('stints').getAll());
  db.close();
  return rows;
}

export async function gradeAgainstPrior(newId) {
  const rows = await allStints();
  const cur = rows.find((r) => r.id === newId);
  if (!cur) return [];
  const prev = rows
    .filter((r) => r.car === cur.car && r.track === cur.track && r.id < newId)
    .sort((a, b) => b.id - a.id)[0];
  if (!prev) return [];

  const newSev = {};
  for (const f of cur.findings) newSev[f.symptom] = f.severity;
  const paceDelta = cur.analysis.median_lap - prev.analysis.median_lap;

  const graded = [];
  const db = await openDb();
  const t = db.transaction('grades', 'readwrite');
  const gs = t.objectStore('grades');
  for (const f of prev.findings) {
    for (const p of f.proposals) {
      let followed = false;
      for (const key of p.keys) {
        const ov = numericValue(prev.setup[key] ?? '');
        const nv = numericValue(cur.setup[key] ?? '');
        if (ov == null || nv == null || nv === ov) continue;
        if ((nv - ov) * p.direction > 0) followed = true;
      }
      if (!followed) continue;
      const sevDelta = (newSev[f.symptom] ?? 0) - f.severity;
      const win = paceDelta < -0.02 || sevDelta < -0.3;
      gs.add({
        created: Date.now() / 1000, car: cur.car, track: cur.track,
        symptom: f.symptom, knob: p.knob, direction: p.direction,
        pace_delta: paceDelta, severity_delta: sevDelta, win: win ? 1 : 0,
        before_stint: prev.id, after_stint: newId,
      });
      graded.push({ symptom: f.symptom, knob: p.knob, pace_delta: paceDelta, severity_delta: sevDelta, win });
    }
  }
  await new Promise((res, rej) => { t.oncomplete = res; t.onerror = () => rej(t.error); });
  db.close();
  return graded;
}

export async function learnedConfidence(car) {
  const db = await openDb();
  const rows = await reqAsync(db.transaction('grades').objectStore('grades').index('car').getAll(car));
  db.close();
  const agg = {};
  for (const r of rows) {
    const key = `${r.symptom}|${r.knob}`;
    const a = agg[key] || (agg[key] = { wins: 0, tries: 0 });
    a.wins += r.win;
    a.tries += 1;
  }
  for (const a of Object.values(agg)) a.conf = (a.wins + 1) / (a.tries + 2);
  return agg;
}

export function applyLearning(findings, conf) {
  for (const f of findings) {
    for (const p of f.proposals) {
      const c = conf[`${f.symptom}|${p.knob}`];
      p.learned = c ? { conf: c.conf, tries: c.tries, wins: c.wins } : { conf: 0.5, tries: 0, wins: 0 };
    }
    f.proposals.sort((a, b) => b.learned.conf - a.learned.conf);
  }
  return findings;
}

export async function knobRecurrence(car, excludeId = null, window = 8) {
  const rows = (await allStints())
    .filter((r) => r.car === car && r.id !== excludeId)
    .sort((a, b) => b.id - a.id)
    .slice(0, window);
  const counts = {};
  for (const r of rows) {
    const seen = new Set();
    for (const f of r.findings) for (const p of f.proposals) seen.add(`${p.knob}|${p.direction}`);
    for (const key of seen) counts[key] = (counts[key] || 0) + 1;
  }
  const out = {};
  for (const [k, v] of Object.entries(counts)) out[k] = { hits: v, stints: rows.length };
  return out;
}

export async function recentStints(limit = 50) {
  return (await allStints())
    .sort((a, b) => b.id - a.id)
    .slice(0, limit)
    .map((r) => ({ id: r.id, created: r.created, file_name: r.file_name, driver: r.driver, car: r.car, track: r.track, stint_num: r.stint_num }));
}

export async function getStint(id) {
  const db = await openDb();
  const row = await reqAsync(db.transaction('stints').objectStore('stints').get(id));
  db.close();
  return row || null;
}
